File: coordinateutils/src/maputils.cpp
#include <pybindings.h>

#include <vector>
#include <math.h>
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <G3Logging.h>
#include <G3Units.h>
#include <G3SkyMap.h>

#include <coordinateutils/CutSkyHealpixMap.h>

#include <coordinateutils/maputils.h>

#include <iostream>

#ifdef OPENMP_FOUND
#include <omp.h>
#endif

using namespace G3Units;

void get_ra_dec_map_cpp(G3SkyMapConstPtr m, G3SkyMapPtr ra, G3SkyMapPtr dec)
{
	ra->EnsureAllocated();
	dec->EnsureAllocated();

	if (!m->IsCompatible(*ra)) {
		log_fatal("Output ra map must match coordinates and dimensions of input sky map");
	}
	if (!m->IsCompatible(*dec)) {
		log_fatal("Output ra map must match coordinates and dimensions of input sky map");
	}

#ifdef OPENMP_FOUND
#pragma omp parallel for
#endif
	for (size_t i = 0; i < m->size(); i++) {
		std::vector<double> radec = m->pixel_to_angle(i);
		(*ra)[i] = radec[0];
		(*dec)[i] = radec[1];
	}

	ra->is_weighted = false;
	ra->units = G3Timestream::None;
	ra->pol_type = G3SkyMap::None;
	dec->is_weighted = false;
	dec->units = G3Timestream::None;
	dec->pol_type = G3SkyMap::None;
}


void reproj_map(G3SkyMapConstPtr in_map, G3SkyMapPtr out_map, int rebin)
{
	out_map->EnsureAllocated();
	if (in_map->coord_ref != out_map->coord_ref) {
		log_fatal("Input and output maps must use the same coordinates");
	}

#ifdef OPENMP_FOUND
#pragma omp parallel for
#endif
	for (size_t i = 0; i < out_map->size(); i++) {
		double val = 0;
		if (rebin > 1) {
			std::vector<double> ra, dec;
			out_map->get_rebin_angles(i, rebin, ra, dec);
			for (size_t j = 0; j < ra.size(); j++) {
				val += in_map->get_interp_value(ra[j], dec[j]);
			}
			val /= ra.size();
		} else {
			std::vector<double> radec = out_map->pixel_to_angle(i);
			val = in_map->get_interp_value(radec[0], radec[1]);
		}
		(*out_map)[i] = val;
	}

	out_map->coord_ref = in_map->coord_ref;
	out_map->is_weighted = in_map->is_weighted;
	out_map->units = in_map->units;
	out_map->pol_type = in_map->pol_type;
}


void reproj_fullsky_healpix_map(std::vector<double> in_map, G3SkyMapPtr out_map,
    bool nest, int rebin)
{
	//grab our out ra dec values
	out_map->EnsureAllocated();

	long nside = npix2nside(in_map.size());
	if (nside < 0) {
		log_fatal("Input map has invalid size for a healpix map");
	}
	HealpixHitPix hitpix(nside, nest, out_map->coord_ref);

#ifdef OPENMP_FOUND
#pragma omp parallel for
#endif
	for (size_t i = 0; i < out_map->size(); i++) {
		double val = 0;
		if (rebin > 1) {
			std::vector<double> ra, dec;
			out_map->get_rebin_angles(i, rebin, ra, dec);
			for (size_t j = 0; j < ra.size(); j++) {
				std::vector<long> pixels;
				std::vector<double> weights;
				hitpix.get_interp_pixels_weights(ra[j], dec[j], pixels, weights, false);
				for (size_t k = 0; k < pixels.size(); k++) {
					val += in_map[pixels[k]] * weights[k];
				}
			}
			val /= ra.size();
		} else {
			std::vector<double> radec = out_map->pixel_to_angle(i);
			std::vector<long> pixels;
			std::vector<double> weights;
			hitpix.get_interp_pixels_weights(radec[0], radec[1], pixels, weights, false);
			for (size_t j = 0; j < pixels.size(); j++) {
				val += in_map[pixels[j]] * weights[j];
			}
		}
		(*out_map)[i] = val;
	}
}


namespace bp = boost::python;
void maputils_pybindings(void){
	bp::def("get_ra_dec_map_cpp", get_ra_dec_map_cpp);
	bp::def("reproj_map", reproj_map,
		(bp::arg("in_map"), bp::arg("out_map"), bp::arg("rebin")=1),
		"Takes the data in in_map and reprojects it onto out_map.  out_map can\n"
		"have a different projection, size, resolution, etc.  Optionally account\n"
		"for sub-pixel structure in the interpolation by setting rebin > 1.");
	bp::def("reproj_fullsky_healpix_map", reproj_fullsky_healpix_map,
		( bp::arg("in_map"), bp::arg("out_map"), bp::arg("nest")=false, bp::arg("rebin")=1),
		"Takes the data in in_map (a full sky healpix map stored as a simple array)\n"
		"and reprojects it onto out_map.  out_map can be any G3SkyMap instance.\n"
		"Optionally account for sub-pixel structure in the interpolation by setting\n"
		"rebin > 1");
}

File: coordinateutils/python/maputils.py
from spt3g import core
from spt3g.coordinateutils import HealpixSkyMap, FlatSkyMap
from spt3g.coordinateutils import reproj_map, get_ra_dec_map_cpp

import numpy as np
import copy

__all__ = [
    'get_ra_dec_map',
    'flatsky_to_healpix',
    'healpix_to_flatsky',
]


@core.usefulfunc
def get_ra_dec_map(map_in):
    """
    Compute the position of each pixel in the map and return maps of ra and dec.

    Parameters:
    -----------
    map_in: G3SkyMap or derived object thereof
        Input map

    Returns:
    --------
    ra, dec: G3SkyMap
        Maps of the coordinates for each pixel, with the same map parameters
        as the input map.
    """

    # store in map objects
    ra = copy.copy(map_in)
    dec = copy.copy(map_in)
    get_ra_dec_map_cpp(map_in, ra, dec)

    return ra, dec


@core.usefulfunc
def healpix_to_flatsky(map_in, nest=False, map_stub=None, rebin=1, interp=False,
                       **kwargs):
    '''
    Re-pixelize a map from Healpix to one of the flat sky projections.

    Parameters:
    -----------
    map_in: numpy array or HealpixSkyMap
        The array containing the input healpix map to reproject.

    nest[False]: bool
        Ordering of the healpix map, if the input is a numpy array.  Ring
        ordering is assumed by default.

    map_stub[None]: FlatSkyMap
        Stub output map object to be used to construct the output map.  If not
        supplied, one will be constructed using the remaining keyword arguments.

    rebin[1]: int
        If supplied and >1, account for sub-pixel structure by integrating
        over a sub-grid on each pixel of the given dimension.  This avoids
        aliasing of power at angular scales beyond the map resolution.

    interp[false]: bool
        If True, use bilinear interpolation to extract values from the input
        map.  Otherwise, the nearest-neighbor value is used.

    All additional keyword arguments are passed to FlatSkyMap to construct
    the output map object.  Required if `map_stub` is not supplied,
    otherwise ignored.

    Returns:
    --------
    output_map: FlatSkyMap
        The reprojected map
    '''

    # Construct the output map
    if map_stub is None:
        if isinstance(map_in, HealpixSkyMap):
            kwargs.setdefault('coord_ref', map_in.coord_ref)
        map_out = FlatSkyMap(**kwargs)
    else:
        if not isinstance(map_stub, FlatSkyMap):
            raise TypeError('Output stub must be a FlatSkyMap')
        map_out = map_stub.Clone(False)

    # Populate output map pixels with interpolation and rebinning
    if not isinstance(map_in, HealpixSkyMap):
        map_in = HealpixSkyMap(map_in, is_nested=nest,
                               coord_ref=map_out.coord_ref,
                               is_weighted=map_out.is_weighted,
                               units=map_out.units,
                               pol_type=map_out.pol_type)
    reproj_map(map_in, map_out, rebin=rebin, interp=interp)

    return map_out


@core.usefulfunc
def flatsky_to_healpix(map_in, map_stub=None, rebin=1, interp=False,
                       fullsky=False, **kwargs):
    '''
    Re-pixelize a map to Healpix from one of the flat projections.

    Parameters:
    -----------
    map_in: FlatSkyMap
        The input map you want to reproject

    map_stub[None]: HealpixSkyMap
        Stub output map object to be used to construct the output map.  If not
        supplied, one will be constructed using the remaining keyword arguments.

    rebin[1]: int
        If supplied and >1, account for sub-pixel structure by integrating
        over a sub-grid on each pixel of the given dimension.  This avoids
        aliasing of power at angular scales beyond the map resolution.

    interp[false]: bool
        If True, use bilinear interpolation to extract values from the input
        map.  Otherwise, the nearest-neighbor value is used.

    fullsky[false]: bool
        If True a full-sky numpy array representation of the map is returned.
        Otherwise, a HealpixSkyMap instance is returned, containing only the
        pixels that overlap with the input map.

    All additional keyword arguments are passed to HealpixSkyMap to construct
    the output map object.  Required if `map_stub` is not supplied,
    otherwise ignored.

    Returns:
    --------
    output_map: numpy array or HealpixSkyMap
        The array containing the healpix map
        If `fullsky` is True, this is a numpy array, otherwise a
        HealpixSkyMap instance.
    '''
    if not isinstance(map_in, FlatSkyMap):
        raise TypeError('Input must be a FlatSkyMap')

    # Construct the output map
    if map_stub is None:
        kwargs.setdefault('coord_ref', map_in.coord_ref)
        map_out = HealpixSkyMap(**kwargs)
    else:
        if not isinstance(map_stub, HealpixSkyMap):
            raise TypeError('Output stub must be a HealpixSkyMap')
        map_out = map_stub.Clone(False)

    # optimize ringsparse storage
    a0 = map_in.alpha_center
    da = map_in.x_res * map_in.shape[1]
    circ = 2 * np.pi * core.G3Units.rad
    amin = np.mod(a0 - da + circ, circ)
    amax = np.mod(a0 + da + circ, circ)
    amin_shift = np.mod(a0 - da + circ + circ / 2, circ)
    amax_shift = np.mod(a0 + da + circ + circ / 2, circ)
    map_out.shift_ra = bool(np.abs(amax - amin) > np.abs(amax_shift - amin_shift))

    # Populate output map pixels with interpolation and rebinning
    reproj_map(map_in, map_out, rebin=rebin, interp=interp)

    if fullsky:
        map_out.dense = True
        return np.asarray(map_out)
    return map_out
